spec.md

```diff
@@ -4695,17 +4698,40 @@ File = {Statement | newline} eof .
 
 Statement = DefStmt | IfStmt | ForStmt | SimpleStmt .
 
-DefStmt = 'def' identifier '(' [Parameters [',']] ')' ':' Suite .
+DefStmt = 'def' identifier '(' [Parameters [',']] ')' ['->' TypeExpr ]  ':' Suite .
 
 Parameters = Parameter {',' Parameter}.
 
-Parameter  = identifier
-           | identifier '=' Test
+Parameter  = identifier [':' TypeExpr] ['=' Test]
            | '*'
-           | '*' identifier
-           | '**' identifier
+           | '*' identifier [':' TypeExpr]
+           | '**' identifier [':' TypeExpr]
            .
 
+LambdaParameters = LambdaParameter {',' LambdaParameter}.
+
+LambdaParameter  = identifier ['=' Test]
+                 | '*'
+                 | '*' identifier
+                 | '**' identifier
+                 .
+           
+TypeExpr = Type {'|' Type}.
```

Comment:
I take it generics of unions, like `List[Noneable[String]]`, will come later? Or like we have in rules' declared providers param, `List[Provider | List[Provider]]`.

For `Noneable`, you could imagine special casing it in the type system if you didn't want union inside generics.

Reply:
I don't follow the comment. With this syntax, `|` is for unions. `List[None | String]` for example. I'm slightly leaning towards this notation (beauty is in the eyes of the beholder, I know).

It would simplify the grammar a lot if I just dropped `|` and go with `List[Noneable[String]]` or `List[Optional[String]]` and `Union[A,B]`.

Reply:
For some reason I had read that grammar as saying that `|` was only permitted at the top level and not in generics. Nevermind.

Reply:
> * Fix TypeArgument to evolve to TypeExpr instead of Type.

Oh, that's probably why I thought that. ;)